### PR TITLE
refactor: bundle `Op` and `Ty` into a `Dialect` struct

### PR DESCRIPTION
**Pull request by alexkeizer**
Morally, we currently identify a dialect with it's `Op` type, which leads to things like 
```lean
abbrev ExceptM  Op [OpSignature Op Ty] := Except (TransformError Ty)
```
Notice how `ExceptM` doesn't actually need `Op`, or its `OpSignature` instance, indeed a comment just above points out:
> Even though we technically only need to know the `Ty` type, our typeclass hierarchy is fully based on `Op`.
> It is thus more convenient to incorporate `Op` in these types, so that there will be no ambiguity errors.

Since we're already pulling in all arguments to `OpSignature` even when we don't need them, it makes sense to bundle `Op` and `Ty` explicitly into a structure 
```lean
structure Dialect where
  Op : Type
  Ty : Type
```
This PR implements this change, implementing one of the alternatives as presented in #250.

The diff is rather huge, but the change is almost completely mechanical, using find/replace.

A notable exception to this is that `deriving Repr` in the definition of `Lets` no longer works, so I've had to manually implement `Repr` instead. Presumably the derivation meta-code knows to insert `[Repr Op] [Repr Ty]` bounds when `Op` and `Ty` are arguments, but it doesn't know to insert `[Repr d.Op]` and `[Repr d.Ty]` bounds when `Lets` has an argument `d : Dialect`.

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
This looks like a good idea. Can you maybe write 1-2 paragraphs of context, potentially even with an example of why this simplifies changes in the future or even reduces the diff of https://github.com/opencompl/ssa/pull/191? I think it is obvious to you, but being explicit here would help me.

AFAIU this pull out a large mechanical change from https://github.com/opencompl/ssa/pull/191, which is perfect as having noisy being 100% mechanical is typically a nice property for a PR.

**Comment by alexkeizer** [comment]
> This looks like a good idea. Can you maybe write 1-2 paragraphs of context, potentially even with an example of why this simplifies changes in the future or even reduces the diff of #191? I think it is obvious to you, but being explicit here would help me.
> 
> AFAIU this pull out a large mechanical change from #191, which is perfect as having noisy being 100% mechanical is typically a nice property for a PR.

You beat me to the punch ;) I was writing a description, but got a bit distracted by #250.
Let's keep discussion of the broader context and trade-offs in #250, I opened this PR mostly because I was curious if anything would break.

File: SSA/Core/Framework.lean [hunk 1]
```diff
@@ -15,19 +15,24 @@ open TyDenote (toType)
...
 abbrev RegionSignature Ty := List (Ctxt Ty × Ty)
 
 structure Signature (Ty : Type) where
   sig     : List Ty
   regSig  : RegionSignature Ty
   outTy   : Ty
 
-class OpSignature (Op : Type) (Ty : outParam (Type)) where
-  signature : Op → Signature Ty
+class OpSignature (d : Dialect) where
```
Comment (goens) [inline SSA/Core/Framework.lean:30]: Shouldn't this be `DialectSignature` then?

**Review by goens — APPROVED** [review]
LGTM (modulo one naming suggestion)

File: SSA/Core/Framework.lean (hunk 1 above)
Comment (alexkeizer) [reply SSA/Core/Framework.lean]: I've also renamed `OpDenote` -> `DialectDenote`

**Comment by bollu** [comment]
@alexkeizer I'm still a little hesitant, given the `simp [(X)]` thing we discovered. I'm happy to merge this, but I wish to note my lingering fear :P

File: SSA/Core/Tactic.lean [hunk 2]
```diff
@@ -120,4 +143,5 @@ macro "simp_peephole" "[" ts: Lean.Parser.Tactic.simpLemma,* "]" "at" Γv:ident
 /-- `simp_peephole` with no extra user defined theorems. -/
 macro "simp_peephole" "at" Γv:ident : tactic => `(tactic| simp_peephole [] at $Γv)
 
+
```
Comment (bollu) [inline SSA/Core/Tactic.lean:146]:
```suggestion
```

**Comment by alexkeizer** [comment]
This now builds locally, so CI should be happy soon as well.

@goens could you give it another look, now that everything is ported over?

**Comment by alexkeizer** [comment]
> @alexkeizer I'm still a little hesitant, given the `simp [(X)]` thing we discovered. I'm happy to merge this, but I wish to note my lingering fear :P

Yeah, some of my optimism has been converted into fear as well :sweat_smile:.
I don't really see an alternative way to reduce verbosity, though, so I'm still cautiously optimistic about going forward.

In particular, since `d : Dialect` is (reducibly) def-eq to `\<d.Op, d.Ty\>` we can always unbundle a particular problematic definition. For example, I defined `Scf.Op` as follows:
```lean
inductive Scf.Op (Op' Ty' : Type) [TyDenote Ty'] [DialectSignature ⟨Op', Ty'⟩] [DialectDenote ⟨Op', Ty'⟩] : Type _
  ...
```

File: SSA/Core/Framework/Dialect.lean [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/-
+Released under Apache 2.0 license as described in the file LICENSE.
+-/
+import SSA.Core.ErasedContext
```
Comment (tobiasgrosser) [inline SSA/Core/Framework/Dialect.lean:4]: Is this import needed?

**Review by tobiasgrosser — APPROVED** [review]
At least code noise wise this seems to go in the right direction.

File: SSA/Core/Tactic.lean [hunk 4]
```diff
@@ -97,15 +97,38 @@ macro "simp_peephole" "[" ts: Lean.Parser.Tactic.simpLemma,* "]" "at" Γv:ident
...
+        use the *default* instance to determine whether a simp-lemma applies to the current goal.
+        Writing `simp [(X)]`, on the other hand, is equivalent to writing `simp [@X _ _ _]`
+          (for as many underscores as `X` takes arguments, implicit or explicit).
+        The parentheses seems to enable `simp` to unify typeclass arguments as well, and thus the
+          simp-lemma applies even for non-standard instances.
+
+        One caveat: `simp [(X)]` only works if `X` is a lemma, *not* if `X` is a definition to be
+        unfolded. Thus, we replace `HVector.denote` with its equation lemmas
+          `(HVector.denote_cons)` and `(HVector.denote_nil)` -/
+        (HVector.denote_cons), (HVector.denote_nil),
```
Comment (goens) [inline SSA/Core/Tactic.lean:131]: wow, that's cursed! thanks for the thorough documentation :)

**Review by goens — APPROVED** [review]
lgtm!

File: SSA/Core/Framework/Dialect.lean (hunk 3 above)
Comment (alexkeizer) [reply SSA/Core/Framework/Dialect.lean:4]: Good call, it is not

**Pull request merged**